core/corehttp/gateway_handler.go

```diff
@@ -956,6 +968,28 @@ func debugStr(path string) string {
 	return q
 }
 
+// Detect 'Cache-Control: only-if-cached' in request and return data if it is already in the local datastore.
+// https://github.com/ipfs/specs/blob/main/http-gateways/PATH_GATEWAY.md#cache-control-request-header
+func (i *gatewayHandler) handleOnlyIfCached(w http.ResponseWriter, r *http.Request, contentPath ipath.Path, logger *zap.SugaredLogger) (requestHandled bool) {
+	if r.Header.Get("Cache-Control") == "only-if-cached" {
+		_, err := i.offlineApi.Block().Stat(r.Context(), contentPath)
+		if err != nil {
+			if r.Method == http.MethodHead {
+				w.WriteHeader(http.StatusPreconditionFailed)
+				return true
+			}
+			errMsg := fmt.Sprintf("%q not in local datastore", contentPath.String())
+			http.Error(w, errMsg, http.StatusPreconditionFailed)
+			return true
+		}
+		if r.Method == http.MethodHead {
+			w.WriteHeader(http.StatusOK)
+			return true
+		}
+	}
+	return false
```

Comment:
You have no happy path for a GET request so it will exit by that point.

In other words.
Let's assume I only store the root block of a file.
You first stat the root block, no error, request is not HEAD and so you return false, and the request continue as usual.
However when the following code tries to actually send the unixfs file, they will hit the network to fetch the childs.

It's not complient with my interpretation of the gateway spec:
> if the gateway already has the data

Well we don't know at that point, we know we have the root but more ? Who knows ?

I think there is two solutions here:
- For now, the spec is just use making things up until it matches whatever our code.
  You can add a `SHOULD` (RFC2119) to the spec sentence accepting that cheap checks might be better.
- At that point `handleOnlyIfCached` return a dagstore object that the rest of the code uses to fetch the data.
  So the rest of the code tries to serve the request as usual but using an offline dagstore.

Sadly I think we cannot do the second option (even tho I like it more) because we cannot send an HTTP error after starting to send the payload (why are we using HTTP again ? browsers .... 😞).

Reply:
@Jorropo thanks for noting this. 

I've made a conscious  decision to only check root block and use `block stat --offline` for everything (even DAG requests) instead of `dag stat --offline` as the latter could act as inexpensive DoS vecor (`dag stat` on Wikipedia DAG does not sound fun, even if all blocks are in the datastore).

The motivation for  `only-if-cached` is to identify gateways which don't have any part of a DAG cached, and prioritize ones that do. 

In that spirit, I agree  that the spec should provide guidance to implementers, who  will have to make a similar decision as I did (or implement custom heuristic/index)

Opened https://github.com/ipfs/specs/pull/297 to clarify spec, but let's merge this PR as-is.

Reply:
@lidel I think that the performance gain `block stat --offline` can be worth it. I'm ok with this just wanted the spec to be in accordance.

I didn't thought you would `dag stat --offline`
But rather do `dagService = offlineBlockservice`

So the directory index code (or file one if that a file, ...) would run in offline mode. It wouldn't be a DOS vector since it's really the same thing as usual but without networking.

Reply:
> Opened https://github.com/ipfs/specs/pull/297 to clarify spec, but let's merge this PR as-is.

:partying_face: